src/manifolds/Stiefel.jl

```diff
@@ -619,3 +620,227 @@ function uniform_distribution(M::Stiefel{n,k,ℝ}, p) where {n,k}
+)
+
+@doc raw"""
+    vector_transport_direction(M::Stiefel, p, X, d, DifferentiatedRetractionVectorTransport{PolarRetraction})
+
+Compute the vector transport by computing the push forward of
+[`retract(::Stiefel, ::Any, ::Any, ::PolarRetraction)`](@ref) Section 3.5 of [^Zhu2017]:
+
+```math
+T_{p,d}^{\text{Pol}}(X) = q*Λ + (I-qq^{\mathrm{T}})X(1+d^\mathrm{T}d)^{-\frac{1}{2}},
+```
+
+where $q = \operatorname{retr}^{\mathrm{Pol}}_p(d)$, and $Λ$ is the unique solution of the Sylvester equation
+
+```math
+    Λ(I+d^\mathrm{T}d)^{\frac{1}{2}} + (I + d^\mathrm{T}d)^{\frac{1}{2}} = q^\mathrm{T}X - X^\mathrm{T}q
+```
+"""
+vector_transport_direction(
+    ::Stiefel,
+    ::Any,
+    ::Any,
+    ::Any,
+    ::DifferentiatedRetractionVectorTransport{PolarRetraction},
+)
+@doc raw"""
+    vector_transport_direction(M::Stiefel, p, X, d, DifferentiatedRetractionVectorTransport{QRRetraction})
+
+Compute the vector transport by computing the push forward of the
+[`retract(::Stiefel, ::Any, ::Any, ::QRRetraction)`](@ref),
+See [^AbsilMahonySepulchre2008], p. 173, or Section 3.5 of [^Zhu2017].
+```math
+T_{p,d}^{\text{QR}}(X) = q*\rho_{\mathrm{s}}(q^\mathrm{T}XR^{-1}) + (I-qq^{\mathrm{T}})XR^{-1},
+```
+where $q = \operatorname{retr}^{\mathrm{QR}}_p(d)$, $R$ is the $R$ factor of the QR
+decomposition of $p + d$, and
+```math
+\bigl( \rho_{\mathrm{s}}(A) \bigr)_{ij}
+= \begin{cases}
+A_{ij}&\text{ if } i > j\\
+0 \text{ if } i = j\\
+-A_{ji} \text{ if } i < j.\\
+\end{cases}
+```
+[^AbsilMahonySepulchre2008]:
+    >Absil, P.-A., Mahony, R. and Sepulchre R.,
+    > _Optimization Algorithms on Matrix Manifolds_
+    > Princeton University Press, 2008,
+    > doi: [10.1515/9781400830244](https://doi.org/10.1515/9781400830244)
+    > [open access](http://press.princeton.edu/chapters/absil/)
+"""
+vector_transport_direction(
+    ::Stiefel,
+    ::Any,
+    ::Any,
+    ::Any,
+    ::DifferentiatedRetractionVectorTransport{QRRetraction},
+)
+
+function vector_transport_direction!(
+    ::Stiefel,
+    Y,
+    p,
+    X,
+    d,
+    ::DifferentiatedRetractionVectorTransport{CayleyRetraction},
+)
+    Pp = I - 1 // 2 * p * p'
+    Wpd = Pp * d * p' - p * d' * Pp
+    WpX = Pp * X * p' - p * X' * Pp
+    q1 = I - 1 // 2 * Wpd
+    return copyto!(Y, (q1 \ WpX) * (q1 \ p))
+end
+
+function vector_transport_direction!(
+    M::Stiefel,
+    Y,
+    p,
+    X,
+    d,
+    ::DifferentiatedRetractionVectorTransport{PolarRetraction},
+)
+    q = retract(M, p, d, PolarRetraction())
+    Iddsqrt = sqrt(I + d' * d)
+    Λ = sylvester(Iddsqrt, Iddsqrt, -q' * X + X' * q)
+    return copyto!(Y, q * Λ + (X - q * (q' * X)) / Iddsqrt)
+end
+function vector_transport_direction!(
+    M::Stiefel,
+    Y,
+    p,
+    X,
+    d,
+    ::DifferentiatedRetractionVectorTransport{QRRetraction},
+)
+    q = retract(M, p, d, QRRetraction())
+    rf = qr(p + d).R
+    return copyto!(Y, q * matUpper2skew((q' * X) / rf) + (X - q * (q' * X)) / rf)
```

Comment:
```suggestion
    rf = qr(p + d).R
    Xrf = X / rf
    qtXrf = q' * Xrf
    return copyto!(Y, q * matUpper2skew(qtXrf) + Xrf - q * qtXrf)
```
I've extracted some common computations. Could you check if that's correct? Additionally, `UpperTriangular(qtXrf) - UpperTriangular(qtXrf)'` is probably be a more standard way of writing `matUpper2skew`.

Reply:
Sure we can also switch to your suggestion for the skew. I wasn't aware of that. For your rewriting – are you sure that holds for non-square matrices, that one can factor our the `q'`?

Reply:
This rewriting relies only on associativity and distributivity of matrix multiplication and this holds for non-square matrices.

By the way, for Stiefel the R factor of QR decomposition is I think an upper triangular matrix so it would be worth applying the `UpperTriangular` wrapper (to dispatch to the more efficient solver for `/`).

Reply:
```
(X - q * (q' * X)) / rf = X / rf - q * (q' * X) / rf
q * (q' * X) / rf = q * (q' * (X / rf))
```

Reply:
I will check for Uppertriangular after writing a small test on Base for the new branch there.